
library(DESeq2)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(wesanderson)
library(stringr)
library(edgeR)
library(pheatmap)
library(ggrepel)
library("RColorBrewer")

# setting the working directory to where the data lies
setwd("C:/Users/jamsh/OneDrive/Desktop/Data_Analysis/Project_1/Melica/Siri_data/")


###

# reading the count table 
countsTable <- read.table("countMatrix.rsem.txt", header= TRUE, row.names=1)
# rounding the counts to integers
countsTableRound <- round(countsTable)


# simplifying the column names (these column names were generated by
# RSEM and they have lots of non-informative words)
cnames <- names(countsTableRound)
for (i in 1:length(names(countsTableRound))){
    cnames[i] = str_replace(cnames[i], "expressions.", "")
      cnames[i] = str_replace(cnames[i], ".genes.results", "")
}
names(countsTableRound) <- cnames


# filtering the rows such that we only keep those that have more than 10
# reads on average
filteredCountsTableRound <- countsTableRound[rowSums(countsTableRound)>10*ncol(countsTableRound),]


##########
### PCA
##########

conds <- read.table("samples.txt", header=TRUE ,stringsAsFactors=TRUE, row.names=1)
dds <-DESeqDataSetFromMatrix(filteredCountsTableRound,
                             colData = conds, 
                             design =~ Genotype)
# TMM normalization
normFactors <- calcNormFactors(filteredCountsTableRound, method="TMM")
N <- colSums(countsTableRound) #vector of library size
tmm.counts <- N*normFactors/exp(mean(log(N*normFactors)))
sizeFactors(dds) <- tmm.counts

# running DESeq
dds <- DESeq(dds)

vsd <- vst(dds, blind=FALSE, nsub=100)


dev.new()
data <- plotPCA(vsd, intgroup=c("Genotype"), returnData=TRUE)
percentVar <- round(100 * attr(data,"percentVar"))

# renaming the labels as T0-M, T0-E, T1-#, ...  
lbs = sub("\\_.*", "", data$name)
lbs = substr(lbs, start=1, stop = 2)
data$Population <- as.factor(lbs)
genotypes <- as.character(data$Genotype)
genotypes[1:5] <- "T0-M"
genotypes[6:8] <- "T0-M"
genotypes[9:11] <- "T0-E"
genotypes[12] <- "T0-M"
genotypes[13:16] <- "T0-E"
genotypes[17:23] <- "T1-E"
genotypes[24:30] <- "T3-M"
genotypes[30:37] <- "T4-M"
data$Genotype <- as.factor(genotypes)


pl <- ggplot(data, aes(PC1,PC2, color=Genotype, shape=Population, label=lbs)) +
    geom_point(size=4, alpha=0.85) +
    xlab(paste0("PC1: ",percentVar[1],"% variance")) +
    ylab(paste0("PC2: ",percentVar[2],"% variance")) +
    theme(panel.background = element_blank(),
          panel.border = element_rect(colour = "black", fill=NA, size=1))

pl + geom_text_repel(aes(label = lbs),
                     point.padding = 0.5,
                     max.overlaps = Inf)

##############
###### DEG
##############

# -------------- FR --------------
# T0 - T1
# --------------------------------

# loading the samples file
conds <- read.table("samples_FRaFRb_T0T1.txt", header=TRUE ,stringsAsFactors=TRUE, row.names=1)

t0t1 <- c("T0.FRa10.1.T0.FRa10.1",
          "T0.FRa6.1.T0.FRa6.1",
          "T0.FRa7.1.T0.FRa7.1",
          "T0.FRa.T0.FRa",
          "T1.FRa7.2.T1.FRa7.2",
          "T1.FRa.T1.FRa",
          "T1.FRb5.2.T1.FRb5.2",
          "T1.FRb6.2.T1.FRb6.2")
sub_countsTableRound <- filteredCountsTableRound[, t0t1]



dds <- DESeqDataSetFromMatrix(countData = sub_countsTableRound, 
                              colData = conds, 
                              design = ~ Genotype)
# TMM
normFactors <- calcNormFactors(sub_countsTableRound, method="TMM")
N <- colSums(sub_countsTableRound) #vector of library size
tmm.counts <- N*normFactors/exp(mean(log(N*normFactors)))
sizeFactors(dds) <- tmm.counts

# test for DEGs
dds <- DESeq(dds, test="Wald")
fdr <- 0.01
# 2**0.58 = 1.5 
# (thresholding FC with 1.5, or equivalently thresholding LFC with 0.58)
res <- results(dds, contrast=c("Genotype","T1","T0"), alpha=fdr, lfcThreshold=0.58)
# thresholding adjusted p-values with 0.01
res <- res[order(res$padj),]
res <- res[!is.na(res$padj),]
degs <- row.names(res[res$padj < fdr,])
LFC <- res[res$padj < fdr,]$log2FoldChange
padj <- res[res$padj < fdr,]$padj
saved_df <- data.frame(deg=degs, LFC=LFC, padj=padj)

norm.counts <- counts(dds, normalized=TRUE)
deg.norm.counts <- norm.counts[degs,]
saved_df <- cbind(saved_df, deg.norm.counts)

path = paste("FRa_FRb/degs_10_consv.txt",sep="")
write.table(saved_df, path, row.names=FALSE, quote=FALSE)


# T0-T3  &  T0-T4
# --------------------------------
conds <- read.table("samples_FRaFRb_T0T3T4.txt", header=TRUE ,stringsAsFactors=TRUE, row.names=1)
t0t3t4 <- c("4.T0.FRa.4.T0.FRa",
            "5.T0.FRa.5.T0.FRa", 
            "6.T0.FRa.6.T0.FRa",
            "7.T0.FRa.7.T0.FRa",
            "8.T0.FRa.8.T0.FRa",
            "T3.FRa2.4.T3.FRa2.4",
            "T3.FRa5.2.T3.FRa5.2",        
            "T3.FRa7.3.T3.FRa7.3",
            "T3.FRa.T3.FRa",   
            "T4.FRa11.4.T4.FRa11.4",
            "T4.FRa14.4.T4.FRa14.4",
            "T4.FRa.T4.FRa",
            "T4.FRb11.3.T4.FRb11.3" 
            )
sub_countsTableRound <- filteredCountsTableRound[, t0t3t4]


# computing DEGs (same procedure as above)
dds <- DESeqDataSetFromMatrix(countData = sub_countsTableRound, 
                              colData = conds, 
                              design = ~ Genotype)
normFactors <- calcNormFactors(sub_countsTableRound, method="TMM")
N <- colSums(sub_countsTableRound) #vector of library size
tmm.counts <- N*normFactors/exp(mean(log(N*normFactors)))
sizeFactors(dds) <- tmm.counts
dds <- DESeq(dds, test="Wald")
norm.counts <- counts(dds, normalized=TRUE)
fdr <- 0.01
for (i in c(3,4)){
    res <- results(dds, contrast = c("Genotype",paste("T",i,sep=""),"T0"), 
                   alpha=fdr, lfcThreshold=0.58)
    res <- res[order(res$padj),]
    res <- res[!is.na(res$padj),]
    degs <- row.names(res[res$padj < fdr,]) 
    LFC <- res[res$padj < fdr,]$log2FoldChange
    padj <- res[res$padj < fdr,]$padj
    saved_df <- data.frame(deg=degs, LFC=LFC, padj=padj)
    
    deg.norm.counts <- norm.counts[degs,]
    saved_df <- cbind(saved_df, deg.norm.counts)
  }


## Heatmap
# -------------------------
dds <- DESeqDataSetFromMatrix(countData = sub_countsTableRound, 
                              colData = conds, 
                              design = ~ Genotype)
normFactors <- calcNormFactors(sub_countsTableRound, method="TMM")
N <- colSums(sub_countsTableRound) #vector of library size
tmm.counts <- N*normFactors/exp(mean(log(N*normFactors)))
sizeFactors(dds) <- tmm.counts
dds <- DESeq(dds)
vsd <- vst(dds, blind=FALSE, nsub=100)
# Taking intersection of DEGs for T0-T3 and T0-T4
degs_3 = read.table("FRa_FRb/degs_30_corrected.txt", header=TRUE, sep=',')
degs_4 = read.table("FRa_FRb/degs_40_corrected.txt", header=TRUE, sep=',')
degs <- intersect(degs_3$gene, degs_4$gene)

mat <- assay(vsd[degs[1:min(length(degs),5000)],])
mat <- assay(vsd)
mat <- mat - rowMeans(mat)

colors = c("#CD9600", "#00A9FF", "#FF61CC")
df <- as.data.frame(colData(dds)[c("Genotype")])
pheatmap(mat, 
         annotation_colors=list(Genotype=c(T0=colors[1],T3=colors[2],T4=colors[3])),
         annotation_col=df, 
         show_rownames = FALSE,
         color=colorRampPalette(c("royalblue3", "ivory2", "brown"))(50),
         filename = "../../../Chapter 1/figures/transcriptome/FR_heatmap.pdf")


# Volcano Plot
library(EnhancedVolcano)


dev.new()
EnhancedVolcano(res,
                lab = rownames(res),
                x = 'log2FoldChange',
                y = 'padj',
                ylab='-Log adjusted P-value',
                title = 'T0-T4',
                pCutoff = 10e-10,
                FCcutoff = 1.5,
                pointSize = 3.0,
                labSize = 2,
                shape = c(1, 4, 23, 25))
                #drawConnectors = FALSE,
                #boxedLabels = TRUE)




# -------------- BI --------------
# T0 - T1
# --------------------------------

t0t1 <- c("T0.39.BId5.1.T0.39.BId5.1",
          "T0.40.BId6.1.T0.40.BId6.1",
          "T0.41.BIe6.1.T0.41.BIe6.1",
          "T1.42.BIe6.2.T1.42.BIe6.2",
          "T1.43.BIe4.2.T1.43.BIe4.2",  
          "T1.44.BId15.T1.44.BId15")
sub_countsTableRound <- filteredCountsTableRound[,t0t1]
conds <- read.table("samples_BIdBIe_T0T1.txt", header=TRUE ,stringsAsFactors=TRUE, row.names=1)

# computing DEGs
dds <- DESeqDataSetFromMatrix(countData = sub_countsTableRound, 
                              colData = conds, 
                              design = ~ Genotype)
normFactors <- calcNormFactors(sub_countsTableRound, method="TMM")
N <- colSums(sub_countsTableRound) #vector of library size
tmm.counts <- N*normFactors/exp(mean(log(N*normFactors)))
sizeFactors(dds) <- tmm.counts
dds <- DESeq(dds, test="Wald")
fdr <- 0.01
res <- results(dds, contrast = c("Genotype","T1","T0"), alpha=fdr)
res <- res[order(res$padj),]
res <- res[!is.na(res$padj),]
degs <- row.names(res[res$padj < fdr,]) 
LFC <- res[res$padj < fdr,]$log2FoldChange
padj <- res[res$padj < fdr,]$padj
saved_df <- data.frame(deg=degs, LFC=LFC, padj=padj)

norm.counts <- counts(dds, normalized=TRUE)
deg.norm.counts <- norm.counts[degs,]
saved_df <- cbind(saved_df, deg.norm.counts)


path = paste("BId_BIe/degs_10_consv.txt",sep="")
write.table(saved_df, path, row.names=FALSE, quote=FALSE)



# T0-T3  &  T0-T4
# --------------------------------
conds <- read.table("samples_BIdBIe_T0T3T4.txt", header=TRUE ,stringsAsFactors=TRUE, row.names=1)

t0t3t4 <- c("T0.10.BIdR.3.T0.10.BIdR.3",
            "T0.11.BIdR.4.T0.11.BIdR.4",
            "T0.12.BIdR.5.T0.12.BIdR.5",  
            "T0.9.BIdR.2.T0.9.BIdR.2",    
            "T3.45.BId2.T3.45.BId2",
            "T3.46.BId9.2.T3.46.BId9.2", 
            "T3.47.BId14.3.T3.47.BId14.3",
            "T4.48.BId9.3.T4.48.BId9.3",
            "T4.49.BIe15.3.T4.49.BIe15.3",
            "T4.50.BIe6.3.T4.50.BIe6.3")
sub_countsTableRound <- filteredCountsTableRound[,t0t3t4]

# computing DEGs
dds <- DESeqDataSetFromMatrix(countData = sub_countsTableRound, 
                              colData = conds, 
                              design = ~ Genotype)
normFactors <- calcNormFactors(sub_countsTableRound, method="TMM")
N <- colSums(sub_countsTableRound) #vector of library size
tmm.counts <- N*normFactors/exp(mean(log(N*normFactors)))
sizeFactors(dds) <- tmm.counts
dds <- DESeq(dds, test="Wald")
fdr <- 0.01
norm.counts <- counts(dds, normalized=TRUE)
for (i in c(4)){
    res <- results(dds, contrast = c("Genotype",paste("T",i,sep=""),"T0"), 
                   alpha=fdr)
    res <- res[order(res$padj),]
    res <- res[!is.na(res$padj),]
    degs <- row.names(res[res$padj < fdr,]) 
    LFC <- res[res$padj < fdr,]$log2FoldChange
    padj <- res[res$padj < fdr,]$padj
    saved_df <- data.frame(deg=degs, LFC=LFC, padj=padj)
    
    deg.norm.counts <- norm.counts[degs,]
    saved_df <- cbind(saved_df, deg.norm.counts)
    
}

## Heatmap
# -------------------------
dds <- DESeqDataSetFromMatrix(countData = sub_countsTableRound, 
                              colData = conds, 
                              design = ~ Genotype)
normFactors <- calcNormFactors(sub_countsTableRound, method="TMM")
N <- colSums(sub_countsTableRound) #vector of library size
tmm.counts <- N*normFactors/exp(mean(log(N*normFactors)))
sizeFactors(dds) <- tmm.counts
dds <- DESeq(dds)
# intersection of T0-T3 & T0-T4 (after removing T0-T1)
degs_3 = read.table("BId_BIe/degs_30_corrected.txt", header=TRUE, sep=',')
degs_4 = read.table("BId_BIe/degs_40_corrected.txt", header=TRUE, sep=',')
degs <- intersect(degs_3$gene, degs_4$gene)

vsd <- vst(dds, blind=FALSE, nsub=100)
mat <- assay(vsd[degs[1:min(length(degs),5000)],])
mat <- mat - rowMeans(mat)

colors = c("#CD9600", "#00A9FF", "#FF61CC")
df <- as.data.frame(colData(dds)[c("Genotype")])
pheatmap(mat, 
         annotation_colors=list(Genotype=c(T0=colors[1],T3=colors[2],T4=colors[3])),
         annotation_col=df, 
         show_rownames = FALSE,
         color=colorRampPalette(c("royalblue3", "ivory2", "brown"))(50),
         filename = "../../../Chapter 1/figures/transcriptome/BI_heatmap.pdf")


###############################
#### Comparison of DEGs
###############################

library(VennDiagram)

# loading DEGs 
degs_30_FR = read.table("FRa_FRb/degs_30_corrected.txt", header=TRUE, sep=',')
degs_40_FR = read.table("FRa_FRb/degs_40_corrected.txt", header=TRUE, sep=',')
degs_30_BI = read.table("BId_BIe/degs_30_corrected.txt", header=TRUE, sep=',')
degs_40_BI = read.table("BId_BIe/degs_40_corrected.txt", header=TRUE, sep=',')

sub1 = degs_30_FR[degs_30_FR$LFC>0,]$gene
sub2 = degs_40_FR[degs_40_FR$LFC>0,]$gene
sub3 = degs_30_BI[degs_30_BI$LFC>0,]$gene
sub4 = degs_40_BI[degs_40_BI$LFC>0,]$gene

ov12 <- intersect(sub1, sub2)
ov13 <- intersect(sub1, sub3)
ov14 <- intersect(sub1, sub4)
ov23 <- intersect(sub2, sub3)
ov24 <- intersect(sub2, sub4)
ov34 <- intersect(sub3, sub4)
ov123 <- intersect(ov12, sub3)
ov124 <- intersect(ov12, sub4)
ov134 <- intersect(ov13, sub4)
ov234 <- intersect(ov23, sub4)
ov1234 <- intersect(ov123, sub4)

#dev.new()
pdf("../../../Chapter 1/figures/transcriptome/venn_fourgroups.pdf")
colors=brewer.pal(n = 4, name = "Set3")
draw.quad.venn(
    area1=length(sub1), 
    area2=length(sub2), 
    area3=length(sub3), 
    area4=length(sub4), 
    n12=length(ov12), 
    n13=length(ov13), 
    n14=length(ov14), 
    n23=length(ov23), 
    n24=length(ov24), 
    n34=length(ov34), 
    n123=length(ov123), 
    n124=length(ov124), 
    n134=length(ov134), 
    n234=length(ov234), 
    n1234=length(ov1234),
    fill=colors,
)
dev.off()


